subcerts.go

```diff
@@ -0,0 +1,508 @@
+}
+
+// InvalidTTL returns true if the credential's validity period is longer than the
+// maximum permitted. This is defined by the certificate's notBefore field
+// (`start`) plus the ValidTime, minus the current time (`now`).
+func (dc *DelegatedCredential) InvalidTTL(start, now time.Time) bool {
+	return dc.ValidTime > (now.Sub(start) + dcMaxTTL).Round(time.Second)
+}
+
+// Validate checks that that the signature is valid, that the credential hasn't
+// expired, and that the TTL is valid. It also checks that certificate can be
+// used for delegation.
+func (dc *DelegatedCredential) Validate(cert *x509.Certificate, vers uint16, now time.Time) (bool, error) {
+	// Check that the cert can delegate.
+	if !canDelegate(cert) {
+		return false, errNoDelegationUsage
+	}
+
+	if dc.IsExpired(cert.NotBefore, now) {
+		return false, errors.New("credential has expired")
+	}
+
+	if dc.InvalidTTL(cert.NotBefore, now) {
+		return false, errors.New("credential TTL is invalid")
+	}
+
+	// Prepare the crednetial for verification.
+	hash := getHash(dc.Scheme)
+	cred := &credential{dc.ValidTime, dc.PublicKey, dc.publicKeyScheme}
+	in, err := prepareDelegation(hash, cred, cert.Raw, dc.Scheme, vers)
+	if err != nil {
+		return false, err
+	}
+
+	// TODO(any) This code overlaps signficantly with verifyHandshakeSignature()
+	// in ../auth.go. This should be refactored.
+	switch dc.Scheme {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+
+		pk, ok := cert.PublicKey.(*ecdsa.PublicKey)
+		if !ok {
+			return false, errors.New("expected ECDSA public key")
+		}
+		sig := new(ecdsaSignature)
+		if _, err = asn1.Unmarshal(dc.Signature, sig); err != nil {
+			return false, err
+		}
+		return ecdsa.Verify(pk, in, sig.R, sig.S), nil
+
+	default:
+		return false, fmt.Errorf(
+			"unsupported signature scheme: 0x%04x", dc.Scheme)
+	}
+}
+
+// Marshal encodes a DelegatedCredential structure per the spec. It also sets
+// dc.Raw to the output as a side effect.
+func (dc *DelegatedCredential) Marshal() ([]byte, error) {
+
+	// The credential.
+	cred := &credential{dc.ValidTime, dc.PublicKey, dc.publicKeyScheme}
+	serialized, err := cred.marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	// The scheme.
+	serializedScheme := make([]byte, 2)
+	binary.BigEndian.PutUint16(serializedScheme, uint16(dc.Scheme))
+	serialized = append(serialized, serializedScheme...)
+
+	// The signature.
+	if len(dc.Signature) > dcMaxSignatureLen {
+		return nil, errors.New("signature is too long")
+	}
+	serializedSignature := make([]byte, 2)
+	binary.BigEndian.PutUint16(serializedSignature, uint16(len(dc.Signature)))
+	serializedSignature = append(serializedSignature, dc.Signature...)
+	serialized = append(serialized, serializedSignature...)
+
+	dc.Raw = serialized
+	return serialized, nil
+}
+
+// unmarshalDelegatedCredential decodes a DelegatedCredential structure.
+func unmarshalDelegatedCredential(serialized []byte) (*DelegatedCredential, error) {
```

Comment:
If unmarshal is un-exported, how do I unmarshal a marshaled DC?

Reply:
You don't, I guess? Originally this function was exported, I'm happy to export this. My thinking was that we wanted to minimize our changes (i.e., additions) to the API as much as possible; since DCs are meant to only be consumed by TLS clients and servers, it made sense to not export it.